### PR TITLE
Dockerize zoobot for pytorch and tensorflow versions

### PR DESCRIPTION
**Pull request by camallen**
I'm leaving this as a draft with no expectation that this is merged. It will have to evolve to be more useful and generic but it's a working start. 

First pass at getting zoobot working in a docker context** along with docker compose.  I've also included some new args for running the model in pytorch to allow finer grained control of the data loading and model training process for pytorch. 

** my mac laptop has no gpu support via docker (or outside of docker) so this is only tested for CPU runs right now. 

Commands to build with docker - can add to the Readme
``` sh
# Pytorch version
docker compose build zoobot
#..... wait for build to finish
docker compose run --rm zoobot bash
# train the model
python train_model_on_catalog.py --experiment-dir results/decals_debug/pytorch --shard-img-size 32 --resize-size 32 --epochs 3 --batch-size 3 --accelerator cpu --gpus 0 --num_data_workers 2 --catalog catalogue.csv

# TF version
docker compose build zoobot_tf
#..... wait for build to finish
# get a bash console up with 
docker compose run --rm zoobot_tf bash
# create shards
python decals_dr5_to_shards.py --labelled-catalog=catalogue.csv --eval-size 11 --shard-dir=data/decals/shards/decals_debug --max-labelled 100 --max-unlabelled=100 --img-size 32
# train etc
python train_model.py --experiment-dir results/decals_debug/tf --shard-img-size 32 --resize-size 32 --train-dir data/decals/shards/decals_debug/train_shards --eval-dir data/decals/shards/decals_debug/eval_shards --epochs 2 --batch-size 1
```

### CODE REVIEW

File: zoobot/pytorch/training/train_with_pytorch_lightning.py [hunk 1]
```diff
@@ -13,7 +13,8 @@
 from zoobot.pytorch.datasets import decals_dr8
 from zoobot.pytorch.training import losses
 from zoobot.pytorch.estimators import define_model
-from zoobot.pytorch.estimators import resnet_detectron2_custom, efficientnet_standard, resnet_torchvision_custom
+# from zoobot.pytorch.estimators import resnet_detectron2_custom, efficientnet_standard, resnet_torchvision_custom
```
Comment (camallen) [inline zoobot/pytorch/training/train_with_pytorch_lightning.py:16]: I removed this only for the unused dependency install...once that is back in i can re import the `resnet_detectron2_custom` properly.

File: Dockerfile.tf [hunk 2]
```diff
@@ -0,0 +1,14 @@
+FROM tensorflow/tensorflow:2.8.0
```
Comment (camallen) [inline Dockerfile.tf:1]: I think this should change to a single base image that can run both systems vs splitting them out...

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,19 @@
+version: '3'
+
+services:
+  zoobot:
+    build:
+      context:  ./
+      dockerfile: Dockerfile
+    volumes:
+      - ./:/usr/src/zoobot
+      - /Users/camallen/workspace/zooniverse/kade/tmp/storage/staging/:/usr/src/zoobot/data/kade/
```
Comment (mwalmsley) [inline docker-compose.yml:10]: Note to self - list as example

File: setup.py [hunk 4]
```diff
@@ -20,5 +20,39 @@
         "Environment :: GPU :: NVIDIA CUDA"
     ],
     packages=setuptools.find_packages(),
-    python_requires=">=3.6"
+    python_requires=">=3.6",
```
Comment (mwalmsley) [inline setup.py:23]: Love the optional dependency stuff

File: zoobot/pytorch/examples/train_model_on_catalog.py [hunk 5]
```diff
@@ -22,14 +22,18 @@
...
+    parser.add_argument('--architecture',
+                        dest='model_architecture', type=str, default='efficientnet')
     parser.add_argument('--epochs', dest='epochs', type=int)
     parser.add_argument('--shard-img-size',
                         dest='shard_img_size', type=int, default=300)
     parser.add_argument('--resize-size', dest='resize_size',
                         type=int, default=224)
     parser.add_argument('--batch-size', dest='batch_size',
                         default=256, type=int)
+    parser.add_argument('--accelerator', type=str, default='gpu')
```
Comment (mwalmsley) [inline zoobot/pytorch/examples/train_model_on_catalog.py:36]: Notes for self

num_data_workers -> num_workers
check effect of auto accel arg here

File: setup.py [hunk 6]
```diff
@@ -20,5 +20,39 @@
...
+            'albumentations',
+            'pyro-ppl == 1.8.0'
+        ],
+        'tensorflow': [
+            'tensorflow >= 2.8',
+            'keras_applications',
+            'tensorflow_probability >= 0.11'
+        ]
+    },
+    install_requires=[
```
Comment (mwalmsley) [inline setup.py:40]: note to self - update readme with install instructions

**Review by mwalmsley — APPROVED** [review]

**Pull request merged**